compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIETransferStridedAccessPattern.cpp

```diff
@@ -0,0 +1,399 @@
+      if (dmaOp.getSourceMemorySpaceAsUInt() == 0) {
+        if (circularDma.getTargetMemorySpaceAsUInt() != 1) {
+          dmaOp.emitOpError() << "has source in L3, but circular dma doesn't "
+                                 "have target in L2.";
+          return failure();
+        }
+
+        // L3 original sizes and strides are needed to create new L2 addressing.
+        SmallVector<OpFoldResult> l3OrigSizes = srcSizes;
+        SmallVector<OpFoldResult> l3OrigStrides = srcStrides;
+
+        size_t dimForCombine = srcSizes.size();
+        if (!isL3AddressingCombinable(srcOffsets, srcSizes, srcStrides,
+                                      dimForCombine)) {
+          dmaOp.emitOpError() << "failed to get dim position for combination";
+          return failure();
+        }
+
+        // Generate L3 side new source offsets/sizes/strides.
+        // Example: [[0, 0, 0] [2, 32, 32] [32, 128, 1]] will become
+        // [[0, 0] [32, 64] [128, 1]] after the first and the innermost dims are
+        // combined.
+        DenseSet<size_t> excludeDims = {dimForCombine};
+        srcOffsets = copyExcludeDims(srcOffsets, excludeDims);
+        srcStrides = copyExcludeDims(srcStrides, excludeDims);
+
+        std::optional<SmallVector<int64_t>> maybeSizes =
+            getConstantIntValues(l3OrigSizes);
+        std::optional<SmallVector<int64_t>> maybeStrides =
+            getConstantIntValues(l3OrigStrides);
+        if (!maybeSizes.has_value() || !maybeSizes.has_value()) {
+          dmaOp.emitOpError()
+              << "failed to get original source sizes / strides.";
+          return failure();
+        }
+        SmallVector<int64_t> sizeVals = maybeSizes.value();
+        SmallVector<int64_t> strideVals = maybeStrides.value();
+
+        int64_t innerDimTotal = strideVals.back() * sizeVals.back();
+        int64_t newInnerSize = sizeVals[dimForCombine] * innerDimTotal;
+
+        size_t lastIndex = l3OrigSizes.size() - 1;
+        excludeDims.insert(lastIndex);
+        srcSizes = copyExcludeDims(srcSizes, excludeDims);
+        srcSizes.push_back(getAsIndexOpFoldResult(ctx, newInnerSize));
+
+        // Generate L2 side new target offsets/sizes/strides.
+        SmallVector<OpFoldResult> newCircularOffsets(l3OrigSizes.size(),
+                                                     rewriter.getIndexAttr(0));
+        tgtCircularOffsets = newCircularOffsets;
+        tgtCircularSizes = copyExcludeDims(l3OrigSizes, excludeDims);
+        tgtCircularSizes.push_back(
+            getAsIndexOpFoldResult(ctx, sizeVals[dimForCombine]));
+        tgtCircularSizes.push_back(getAsIndexOpFoldResult(ctx, innerDimTotal));
+        tgtCircularStrides = getNewL2Strides(sizeVals);
+        tgtCircularStrides.insert(
+            tgtCircularStrides.begin() + dimForCombine,
+            getAsIndexOpFoldResult(ctx, strideVals[dimForCombine]));
+      }
+
+      // Generate new L3 target addressing, and new L2 source addressing.
```

Comment:
For this as well as the preceding L3-source-L2-target, can you make one function to which you can send in the corresponding datas/output vectors to modify ?

Would make the code cleaner.
The first check of source-memory-as-int vs target-memory-as-int you may do it outside but I see that major portion of the code can be in fact decoupled out of this function and be reused.

Reply:
Thanks for the suggestion! I refactored the codes.